proctor-consumer/src/main/java/com/indeed/proctor/consumer/ProctorConsumerUtils.java

```diff
@@ -106,16 +108,26 @@ public static String getForceGroupsStringFromRequest(@Nonnull final HttpServletR
             return "";
         }
 
-        for (int i = 0; i < cookies.length; i++) {
-            if (FORCE_GROUPS_COOKIE_NAME.equals(cookies[i].getName())) {
-                final String cookieValue = cookies[i].getValue();
-                return Strings.nullToEmpty(cookieValue);
-            }
-        }
-
-        return "";
+        return Arrays.stream(cookies).sequential()
+                .filter(Objects::nonNull)
+                .filter(x -> FORCE_GROUPS_COOKIE_NAME.equals(x.getName()))
+                .sorted(COOKIE_COMPARATOR)
+                .map(Cookie::getValue)
+                .filter(Objects::nonNull)
+                .collect(Collectors.joining(","));
     }
```

Comment:
I now realized that the current logic (taking force-group options from one of parameter, header and cookie) might result in a confusing behavior because forceGroups options in cookie are ignored when the request has force-groups in query or header. For example,
1. Access foo.example.com?prforceGroups=a_tst1 with cookie prforceGroups=b_tst2 on example.com. => b_tst2 is ignored.
2. The response includes a Set-Cookie header for a_tst1 on foo.example.com.
3. The opened page calls an API example.com/api/foo with cookies prforceGroups=b_tst2 and prforceGroups=a_tst1. => Both a_tst1 and b_tst2 are forced.
4. The response includes a Set-Cookie header for both a_tst1 and b_tst2 on foo.example.com.

It's fine to keep it as it is, but I'm concerned that this problem becomes more serious by introducing this feature.

Reply:
I don't think it's confusing for people to get the cookie ignored by the parameter (specifically because it overwrites the cookie).

I don't think this example works though. In step 3, a request to example.com/api/foo will only get the prforceGroup cookie for b_tst2. It can't see the cookies on foo.example.com. So you don't get a_tst1 and b_tst2 forced on example.com.

The root problem is probably that people are writing cookies at ".example.com" instead of "www.example.com". Ideally, you should only ever have one force group cookie applicable. But that's not the case.

There's also a side problem where we launder cookies; we will re-write the force groups cookie, even if we read it on this very request. Let's say we have two applications that are both mounted on www.example.com: HomepageWebapp and SearchWebapp. Let's say for some reason HomepageWebapp sets the forced group cookie on "www.example.com" and SearchWebapp sets the cookie on ".example.com".

* Visit Homepage with "?prforceGroups=a_tst1". We get a cookie on "www.example.com" with value = "a_tst1".
* Visit SearchWebapp with "?prforceGroups=b_tst2". We get a cookie on ".example.com" with value = "b_tst2".
* Visit SearchWebapp again _without a prforceGroups parameter_. Because long paths get sorted first in the "Cookie" header, we will end up using the cookie from "www.example.com" which is older instead of the one from ".example.com" which is newer. In other words, they'll get a forced value for "a_tst" and not for "b_tst". And that's almost certainly not what they're expecting.
If they are forcing the same test in the different apps, nothing we do here will help them. But we can at least recognize this situation and give them "a_tst1,b_tst2" which is closer!

Reply:
> I don't think this example works though. In step 3, a request to example.com/api/foo will only get the prforceGroup cookie for b_tst2. It can't see the cookies on foo.example.com. So you don't get a_tst1 and b_tst2 forced on example.com. (Unless it's foo.example.com/api/foo and it's setting cookies on ".example.com")

Sorry, I made a typo in the step 3 in the example. As you mentioned it should be `foo.example.com/api/foo`.

Reply:
> So my concern is that prforceGroups cookies are not used if a request has prforceGroups query parameter or header.
A valid concern. But I think people by now are used to this behavior and expect that a header will override the groups cookie. The problem is that the groups cookie that's chosen may not be the cookie they created most recently.

(Maybe there's a future enhancement where we encode a last-updated-date in the prforceGroups cookie and only use the latest one? We'd need to avoid re-upping that date when merely reading from the cookie and re-writing it though.)

Reply:
Got it. Let's consider improving this when users actually get confused about this.